osu.Game.Rulesets.Osu/Edit/OsuSelectionHandler.cs

```diff
@@ -207,28 +192,118 @@ public override bool HandleRotation(float delta)
+
+                // guard against no-ops and NaN.
+                if (scale.X != 0 && selectionQuad.Width > 0)
+                    newPosition.X = selectionQuad.TopLeft.X + xOffset + (h.X - selectionQuad.TopLeft.X) / selectionQuad.Width * (selectionQuad.Width + scale.X);
+
+                if (scale.Y != 0 && selectionQuad.Height > 0)
+                    newPosition.Y = selectionQuad.TopLeft.Y + yOffset + (h.Y - selectionQuad.TopLeft.Y) / selectionQuad.Height * (selectionQuad.Height + scale.Y);
+
+                h.Position = newPosition;
+            }
+
+            return true;
+        }
+
+        private (bool X, bool Y) isQuadInBounds(Quad quad)
+        {
+            bool xInBounds = (quad.TopLeft.X >= 0) && (quad.BottomRight.X <= DrawWidth);
+            bool yInBounds = (quad.TopLeft.Y >= 0) && (quad.BottomRight.Y <= DrawHeight);
+
+            return (xInBounds, yInBounds);
+        }
+
+        private void moveSelectionInBounds()
         {
             var hitObjects = selectedMovableObjects;
 
             Quad quad = getSurroundingQuad(hitObjects);
+            Vector2 delta = new Vector2(0);
 
-            Vector2 newTopLeft = quad.TopLeft + delta;
-            if (newTopLeft.X < 0)
-                delta.X -= newTopLeft.X;
-            if (newTopLeft.Y < 0)
-                delta.Y -= newTopLeft.Y;
+            if (quad.TopLeft.X < 0)
+                delta.X -= quad.TopLeft.X;
+            if (quad.TopLeft.Y < 0)
+                delta.Y -= quad.TopLeft.Y;
 
-            Vector2 newBottomRight = quad.BottomRight + delta;
-            if (newBottomRight.X > DrawWidth)
-                delta.X -= newBottomRight.X - DrawWidth;
-            if (newBottomRight.Y > DrawHeight)
-                delta.Y -= newBottomRight.Y - DrawHeight;
+            if (quad.BottomRight.X > DrawWidth)
+                delta.X -= quad.BottomRight.X - DrawWidth;
+            if (quad.BottomRight.Y > DrawHeight)
+                delta.Y -= quad.BottomRight.Y - DrawHeight;
 
             foreach (var h in hitObjects)
                 h.Position += delta;
+        }
 
-            return true;
+        /// <summary>
+        /// Clamp scale where selection does not exceed playfield bounds or flip.
+        /// </summary>
+        /// <param name="hitObjects">The hitobjects to be scaled</param>
+        /// <param name="reference">The anchor from which the scale operation is performed</param>
+        /// <param name="scale">The scale to be clamped</param>
+        /// <returns>The clamped scale vector</returns>
+        private Vector2 getClampedScale(OsuHitObject[] hitObjects, Anchor reference, Vector2 scale)
+        {
+            float xOffset = ((reference & Anchor.x0) > 0) ? -scale.X : 0;
+            float yOffset = ((reference & Anchor.y0) > 0) ? -scale.Y : 0;
+
+            Quad selectionQuad = getSurroundingQuad(hitObjects);
+
+            //todo: this is not always correct for selections involving sliders
```

Comment:
Do you mind elaborating why in this same comment, so that future travelers don't have to wonder?